src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,25 +1,78 @@
 package com.es.phoneshop.model.product;
 
-import java.util.List;
+import com.es.phoneshop.exceptions.NoProductWithSuchIdException;
+
+import java.math.BigDecimal;
+import java.util.*;
+import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
+    private final List<Product> products = new ArrayList<>();
+    private volatile long maxId;
+
+    private final Object mutex = new Object();
```

Comment:
Can we simply use synchronized methods instead of mutex object?

Reply:
I've replaced some synchronized statements by synchronized methods. Also I've removed mutex object, now synchronized statements use products.